model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientModelCriteriaBuilder.java

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.jpa.client.storage;
+
+import org.keycloak.models.map.jpa.client.entity.JpaClientEntity;
+import org.keycloak.models.map.jpa.client.entity.JpaClientAttributeEntity;
+import java.util.Arrays;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientModel.SearchableFields;
+import org.keycloak.models.map.jpa.hibernate.jsonb.JsonbType;
+import org.keycloak.models.map.storage.CriterionNotSupportedException;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.storage.SearchableModelField;
+
+public class JpaClientModelCriteriaBuilder implements ModelCriteriaBuilder<ClientModel, JpaClientModelCriteriaBuilder> {
+
+    private BiFunction<CriteriaBuilder, Root<JpaClientEntity>, Predicate> predicateFunc = null;
+
+    public JpaClientModelCriteriaBuilder() {
+    }
+
+    private JpaClientModelCriteriaBuilder(BiFunction<CriteriaBuilder, Root<JpaClientEntity>, Predicate> predicateFunc) {
+        this.predicateFunc = predicateFunc;
+    }
+
+    private void validateValue(Object[] value, int expectLength, SearchableModelField field, Operator op) {
+        if (value == null || value.length != expectLength) {
+            throw new CriterionNotSupportedException(field, op, "Invalid argument: " + Arrays.toString(value));
+        }
+    }
+
+    @Override
+    public JpaClientModelCriteriaBuilder compare(SearchableModelField<? super ClientModel> modelField, Operator op, Object... value) {
+        switch (op) {
+            case EQ:
+                if (modelField.equals(SearchableFields.REALM_ID) || 
+                    modelField.equals(SearchableFields.CLIENT_ID) ||
+                    modelField.equals(SearchableFields.ENABLED)) {
+
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.SCOPE_MAPPING_ROLE)) {
+                    validateValue(value, 1, modelField, op);
+
+                    return new JpaClientModelCriteriaBuilder((cb, root) -> 
+                        cb.isTrue(cb.function("@>",
+                            Boolean.TYPE,
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fScopeMappings")),
+                            cb.literal("\"" + value[0] + "\"")))
```

Comment:
(SQL injection) Can `value[0]` contain `"`?

Reply:
Right now it cannot, because we call searchable field `SCOPE_MAPPING_ROLE` only from `MapClientProvider.preRemove`: https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java#L354

And there is `RoleModel.getId()` used as value.

But I tried to test it to be on safe site and it seems it should be ok (if I understand it correctly). I tried to use `.compare(SearchableFields.SCOPE_MAPPING_ROLE, Operator.EQ, "\"::jsonb=true;drop schema public cascade;")` and the result was

```
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type json
  Detail: Token ""::jsonb=true;drop schema public cascade;" is invalid.
  Where: JSON data, line 1: "::jsonb=true;drop schema public cascade;
```

Reply:
Then I don't understand there is the `"\""` in the `literal`. Should it be without?

Reply:
Without it, it throws PSQLException:

```
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type json
  Detail: Token "c3a7d887" is invalid.
  Where: JSON data, line 1: c3a7d887...
```